### DISCUSSION TITLE
Difficulty with unflattening a model (RMG016, RMG020)

### DISCUSSION THREAD

**Discussion by sander1095** [riok/mapperly#1323] [Q&A]
Hello!

I'm trying to map the following but this doesn't work. I can't really understand why.

```csharp
// Source
public class Order
{
    public required int Id { get; set; }

    public required string CustomerId { get; set; }
}

// Target
public class OrderDto
{
    public required int Id { get; set; }
    public required OrderCustomerDto Customer { get; set; }
}

public class OrderCustomerDto
{
    public required string Id { get; set; }
}

[Mapper]
public static partial class OrderDtoMapper
{
    [MapProperty("CustomerId", "Customer.Id")]
    public static partial OrderDto ToResponseModel(this Order order);
}
```

I get the following errors:

- RMG016 Cannot map to init only member path Uitgekookt.Orderpick.API.Models.BoxScanner.OrderCustomerDto.Customer.Id
- RMG020 The member CustomerId on the mapping source type Order is not mapped to any member on the mapping target type OrderDto.

If I remove the `required` keyword from `public required OrderCustomerDto Customer { get; set; }`, the errors are gone, but the generated code will generate a nullref:

```csharp
// namespace info has been removed for privacy reasons
public static partial OrderDto ToResponseModel(this Order order)
{
    var target = new OrderDto()
    {
        Id = order.Id,
    };
    target.Customer.Id = order.CustomerId;
    return target;
}
```

I also think the MapProperty should not be necessary because doesn't mapperly use pascalcase to understand mappings?

To summarize:

- Why does this not work?
- What needs to be done to make this work?

Thanks! <3

**Comment by latonz** [comment]
> Why does this not work?

Unfortunately unflattening to init/required properties is not supported atm: https://github.com/riok/mapperly/issues/643.

> What needs to be done to make this work?

A PR implementing this feature, feel free to contribute 😉
